plugins/catalog-backend/src/experiments/processing.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, LocationEntity } from '@backstage/catalog-model';
+
+export type LocationReader = {
+  read(location: string): Buffer;
+};
+
+export type Relation = {
+  type: string;
+  source: string;
+  target: string;
+};
+
+export type ProcessingResult =
+  | {
+      type: 'error';
+      error: {
+        name: string;
+        message: string;
+      };
+    }
+  | {
+      type: 'entity';
+      entity: Entity;
+    }
+  | {
+      type: 'relation';
+      relation: Relation;
+    };
+
+export type ProcessingContext = LocationReader & {
+  emit(result: ProcessingResult): void;
+};
+
+export type EntityProcessor = {
+  process(entity: Entity, context: ProcessingContext): Entity;
```

Comment:
All processing is done on entities. We skip data, and location reading gets moved over to the `reader`